### PR TITLE
V2.0.0

### PR DESCRIPTION
**Pull request by oniatus**
Internal PR to merge v2.0.0 back into develop.

Not ready for merge yet, I will try to include the behavior changes from #3023 first.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Build triggered. sha1 is merged.

**Comment by GooeyHub** [comment]
Build finished. 
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Build triggered. sha1 is original commit.

**Comment by GooeyHub** [comment]
Build started sha1 is original commit.
Uh oh, something went wrong with the build. Need to check on that

**Comment by oniatus** [comment]
Included the behavior PR here. Reminder for the outstanding PRs in modules which depend on this one: 
- https://github.com/Terasology/Pathfinding/pull/34
- https://github.com/Terasology/Behaviors/pull/7
- https://github.com/Terasology/WildAnimals/pull/22
- https://github.com/Terasology/GooeysQuests/pull/15 
- https://github.com/Terasology/LightAndShadow/pull/59
- https://github.com/Terasology/LightAndShadowResources/pull/21
- https://github.com/Terasology/StaticCities/pull/5

**Comment by GooeyHub** [comment]
Build finished. 
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Build triggered. sha1 is merged.

**Comment by GooeyHub** [comment]
Build started sha1 is merged.
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Build finished. 
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Build triggered. sha1 is merged.

**Comment by GooeyHub** [comment]
Build started sha1 is merged.
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Build finished. 
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Build triggered. sha1 is merged.

**Comment by GooeyHub** [comment]
Build started sha1 is merged.
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Build finished. 
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
In addition to the module PR list in an earlier comment I'm also adding three new related modules to the lineup:

* https://github.com/Terasology/Behaviors (linked in the prior comment but not yet in the lineup)
* https://github.com/Terasology/TutorialBehaviors (also part of the GSOC project for BTs)
* https://github.com/Terasology/TutorialSectors (Sectors naturally)

Conflicts were surprisingly benign considering how substantial the two GSOC projects are, especially when sitting in a branch for months first :-)

Additional changes I had to make:

* https://github.com/Terasology/Pathfinding/commit/6e52374b03282aa5e4f68cb75200e3323f8e4544 introduced a new node during the BT project, but wasn't converted to the new system. I commented out the whole class, need @dkambersky and @arpan98 to help update it to the new system
* Zones introduced `getZones` and `getNamedZone` to the `WorldGenerator` interface which breaks every existing world in unrelated modules since those two aren't implemented. I put a minimal default implementation in the interface, need to consider whether that's a good option or if other world generators should be updated (I suspect most would just end up with a dummy implementation anyway). Those two methods could also use some javadoc, wink wink @Vizaxo :D
* The Behaviors module had a unit test `ClusterDebugger` originally written long ago by @synopia that relied on a headless environment since changed for the shiny new `ModuleTestingEnvironment` or something similar - likely of interest to @oniatus and/or @kaen. I just commented out one line so the test will likely fail, but it now compiles!
* https://github.com/Terasology/AdvancedBehaviors ... is pretty broken. IIRC it came up and became a thing during GSOC as @nihal111 added a bunch of new behaviors and there was some uncertainty about what to do with it? I'm not sure if similar logic was introduced in the Behaviors module thus superseding AB or if we need to update it with a nice PR like the other modules. 
  *  Also pinging @dkambersky and @flo on that one, pretty sure we talked about it once or more. Not fixing right away, which will probably break some other stuff and hold up testing for a bit. Was it the sort of controversial change where Nihal had made some new cool easy to use behaviors built on the old systems, which would be a bit more complicated to approach in the new system, even if longer term it would be more powerful?
  * Seems the main other thing broken is actually just WildAnimalsGenome - indirectly through WildAnimals, which has the dependency on AB but doesn't appear to use anything directly from it (huh). I guess maybe more accurately WAG should depend on AB.
* On testing Light & Shadow the creatures are *somewhat* walking again, but on opening the BT editor with F5 my game crashed with an NPE at `org.terasology.rendering.nui.TextLineBuilder.addText(TextLineBuilder.java:56)` - same seems to happen even just running TutorialBehaviors

Overall the game seems surprisingly stable with this merged locally in a mega-workspace! Merging this and continuing to organize some module stuff :-)

Thanks again to @dkambersky and @Vizaxo for their GSOC items and to all the mentors and other related helpers!

**Comment by Cervator** [comment]
... I think that's actually it? That was a lot of juggling in module land, but I think everything is merged and built :-)

AdditionalBehaviors is the one thing that's broken, which then indirectly impacts other stuff. Everything else remains green (or in Jenkins-speak: blue)

@dkambersky: halp :D

**Comment by dkambersky** [comment]
To sorta recap the bunch of discussions we had around AdvancedBehaviors, the new behavior system etc:

AdvancedBehaviors was created for the old system, mainly to fix a major issue with it (the tree not re-evaluating when continuously running a long-running node, causing the entity to take way too long to respond to outside events in many cases).

The new system doesn't have that issue, as on every tick, the tree is evaluated from the root, so any changes are picked up instantly (mainly using DynamicSelector vs the normal Selector), so AdvancedBehaviors isn't needed.  I wrote about that a bit [here](https://dkambersky.github.io/tera/2017/07/07/on-responsiveness.html)

Now, AdvancedBehaviors had a different useful part to it too, and that was the 'add component to add behavior driven by a system' functionality. That's not present in the new system, in part because the AdvBeh approach assumes every behavior has a strict, rigid integer priority, behaviors are switched out as a whole so any progress/state is lost between transitions, and it limits expressivity a lot - more complex loop/switch/control flow logic is practically impossible with just the 'integer priority swaps out trees' approach - in the new system, every entity's behavior based on a tree, making it more expressive (all the BT logic can be used) and more maintainable - all the behavior is there in one place.

We discussed this and agreed a bunch of times in #ai on Slack, but the logs are gone, so I'm just writing the reasoning down for posterity.

**TL;DR:** AdvancedBehaviors in its current state is not needed, as the new Behavior system fixed the problems in the old system that AdvancedBehaviors was working around.

I'll have a look at the CheckTargetSet node when I can. There are some issues with the Work system, which the L&S creatures rely on, I've gotten some way to fixing it over the break but I don't have much time to work on it now - I'll try at least publishing my commits somewhere, in the relevant repos.
